dev/benchmarks/macrobenchmarks/test/util.dart

```diff
@@ -0,0 +1,221 @@
+''';
+
+void macroPerfTestE2E(
+  String testName,
+  String routeName, {
+  Duration pageDelay,
+  Duration duration = const Duration(seconds: 3),
+  Duration timeout = const Duration(seconds: 30),
+  DriveCallback body,
+  DriveCallback setup,
+}) {
+  assert(() {
+    debugPrint(_kDebugWarning);
+    return true;
+  }());
+  final WidgetsBinding _binding = E2EWidgetsFlutterBinding.ensureInitialized();
+  assert(_binding is E2EWidgetsFlutterBinding);
+  final E2EWidgetsFlutterBinding binding = _binding as E2EWidgetsFlutterBinding;
+  binding.framePolicy = LiveTestWidgetsFlutterBindingFramePolicy.benchmarkLive;
+
+  testWidgets(testName, (WidgetTester tester) async {
+    assert((tester.binding as LiveTestWidgetsFlutterBinding).framePolicy ==
+        LiveTestWidgetsFlutterBindingFramePolicy.fullyLive);
+    app.main();
+    await tester.pumpAndSettle();
+
+    // The slight initial delay avoids starting the timing during a
+    // period of increased load on the device. Without this delay, the
+    // benchmark has greater noise.
+    // See: https://github.com/flutter/flutter/issues/19434
+    await tester.binding.delayed(const Duration(microseconds: 250));
+
+    final Finder scrollable =
+        find.byKey(const ValueKey<String>(kScrollableName));
+    expect(scrollable, findsOneWidget);
+    final Finder button =
+        find.byKey(ValueKey<String>(routeName), skipOffstage: false);
+    await tester.ensureVisible(button);
+    expect(button, findsOneWidget);
+    await tester.pumpAndSettle();
+    await tester.tap(button);
+
+    if (pageDelay != null) {
+      // Wait for the page to load
+      await tester.binding.delayed(pageDelay);
+    }
+
+    if (setup != null) {
+      await setup(tester);
+    }
+
+    await watchPerformance(binding, () async {
+      final Future<void> durationFuture = tester.binding.delayed(duration);
+      if (body != null) {
+        await body(tester);
+      }
+      await durationFuture;
+    });
+  }, semanticsEnabled: false, timeout: Timeout(timeout));
+}
+
+Future<void> watchPerformance(
+  E2EWidgetsFlutterBinding binding,
+  Future<void> action(),
+) async {
+  // This method might be good as part of e2e,
+  // so is the helper `PerformanceWatcher` class.
+  final FrameTimingSummarizer frameTimes = FrameTimingSummarizer();
+  final TimingsCallback watcher = frameTimes.addData;
+  binding.addTimingsCallback(watcher);
+  await action();
+  binding.removeTimingsCallback(watcher);
+  // TODO(CareF): determine if it's running on firebase and report metric online
+  binding.reportData = <String, dynamic>{'performance': frameTimes.summary};
+}
+
+/// This class records [FrameTiming] and summarizes the building statistics.
+///
+/// Without otherwise noticed, all time in this class is in unit microseconds.
+class FrameTimingSummarizer {
+  FrameTimingSummarizer({List<FrameTiming> initData})
+      : _data = List<FrameTiming>.from(initData) ?? <FrameTiming>[];
```

Comment:
As you pointed out in https://github.com/flutter/flutter/pull/61509#discussion_r457753936, most people who use this function probably don't expect the `List<FrameTiming>` to be mutating in their first thought.

I'd suggest to let your `watcher` record a `List<FrameTiming>`, and give that final list to `FrameTimingSummarizer` after that list is fully recorded.

You can then extract the build and raster duration and sort in the constructor. I feel it's probably still cleaner to call `_findPercentile` in each getter so you don't have to introduce a lot of class member fields.

Reply:
Done with refactoring `FrameTimingSummarizer` using factory constructor.